app/controllers/v1/custom_fields_controller.rb

```diff
@@ -0,0 +1,63 @@
+module V1
+  class CustomFieldsController < ApplicationController
+    before_action :valid_vocabulary?, :field_key_present?
+
+    def create
+      if vocabulary.custom_fields.key?(params[:field_key])
+        render json: URIService::JSON.errors('Field key already exists'), status: 409
+      else
+        vocabulary.add_custom_field(create_params)
+
+        if vocabulary.save
+          render json: URIService::JSON.custom_field(vocabulary, params[:field_key]), status: 201
+        else
+          render json: URIService::JSON.errors(vocabulary.errors.full_messages), status: 400
+        end
+      end
+    end
+
+    def update
+      if vocabulary.custom_fields.key?(params[:field_key])
+        vocabulary.update_custom_field(update_params)
+
+        if vocabulary.save
+          render json: URIService::JSON.custom_field(vocabulary, params[:field_key]), status: 200
+        else
+          render json: URIService::JSON.errors(vocabulary.errors.full_messages), status: 400
+        end
+      else
+        render json: URIService::JSON.errors('Not Found'), status: 404
+      end
+    end
+
+    def destroy
+      if vocabulary.custom_fields.key?(params[:field_key])
+        vocabulary.delete_custom_field(params[:field_key])
+
+        if vocabulary.save
+          head :no_content
+        elsif
+          render json: URIService::JSON.errors(vocabulary.errors.full_messages), status: 400
+        end
+      else
+        render json: URIService::JSON.errors('Not Found'), status: 404
+      end
+    end
+
+    private
+
+      def field_key_present?
+        if params.fetch(:field_key, nil).blank?
+          render json: URIService::JSON.errors('Field key must be present.'), status: 400
+        end
+      end
+
+      def create_params
+        params.permit(:field_key, :label, :data_type)
+      end
+
+      def update_params
+        params.permit(:field_key, :label)
```

Comment:
Do we want to allow people to update field keys, or is it better to just delete an old field and add a new one with a different field key?  We don't allow people to change the data_type of an existing field because that could cause issues with currently-stored data, but I feel like changing the field_key would be almost as problematic.  (I could be convinced otherwise though!  Just wanted to talk this through.)

Reply:
field_keys can't be changed with an update operation, the update operation needs to know the field key so it knows which one to update.

Reply:
It is confusing tho, because its not a model itself i am doing it a little bit differently.

Reply:
Oh right, I wasn't considering the context.  I was just thinking this as if it were a normal whitelist scenario.